### PR TITLE
Add address checking in process_load

### PR DESCRIPTION
**Pull request by badochov**
JIRA: RTOS-529

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: proc/process.c [hunk 1]
```diff
@@ -395,26 +395,186 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
 }
 
 
+static int process_validatePtr(void *map_start, size_t map_length, void *ptr_start, size_t ptr_size)
```
Comment (agkaminski) [inline proc/process.c:398]: Remember about camelCase ;)
```suggestion
static int process_validatePtr(void *mapStart, size_t mapLength, void *ptrStart, size_t ptrSize)
```

File: proc/process.c [hunk 2]
```diff
@@ -395,26 +395,186 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
...
+{
+	void *map_end = ((char *)map_start) + map_length;
+	void *ptr_end = ((char *)ptr_start) + ptr_size;
+
+	return ((ptr_size == 0) ||
```
Comment (agkaminski) [inline proc/process.c:403]: Take a look here, this is two-person team researched optimal overlap check algorithm:
https://github.com/phoenix-rtos/plo/blob/486d7526322c6d835434d0b567dbc908c5522b35/syspage.c#L223

Comment (agkaminski) [reply proc/process.c:403]: So:
`((ptr_end > map_start) && (ptr_start < map_end) && (ptr_start < ptr_end)))`
should be enough

File: proc/process.c [hunk 3]
```diff
@@ -395,26 +395,186 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
...
 	offs_t offs;
 	char *snameTab;
 
-	shdr = (void *)((char *)ehdr + ehdr->e_shoff);
-	shdr += ehdr->e_shstrndx;
-	snameTab = (char *)ehdr + shdr->sh_offset;
+	ret = process_loadValidateElf32(iehdr, size);
+	if (ret < 0) {
+		return ret;
+	}
```
Comment (agkaminski) [inline proc/process.c:571]: Imho we don't really need retval of process_loadValidateElf32, as it either return -ENOEXEC or 0
```suggestion
	if (process_loadValidateElf32(iehdr, size) != 0) {
		return -ENOEXEC;
	}
```

File: proc/process.c (hunk 2 above)
Comment (badochov) [reply proc/process.c]: but here it's not an overlap check but a check if the map includes the pointer

File: proc/process.c [hunk 4]
```diff
@@ -395,26 +395,186 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
...
+static int process_validatePtr(void *map_start, size_t map_length, void *ptr_start, size_t ptr_size)
+{
+	void *map_end = ((char *)map_start) + map_length;
+	void *ptr_end = ((char *)ptr_start) + ptr_size;
+
+	return ((ptr_size == 0) ||
+			   ((ptr_end > map_start) && (ptr_end <= map_end) &&
+				   (ptr_start >= map_start) && (ptr_start < ptr_end))) ?
+        1 :
+        0;
```
[the remark this answers is not in the document]
Comment (badochov) [reply proc/process.c]: I renamed the function to isPtrValid

[the remark this answers is not in the document]
Comment (agkaminski) [reply proc/process.c]: I tend to just ignore it's nonsense, but it's ok to turn it off

File: proc/process.c [hunk 5]
```diff
@@ -395,26 +395,185 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
...
+	return ((ptrSize == 0) ||
+			((ptrEnd > mapStart) && (ptrEnd <= mapEnd) &&
+				(ptrStart >= mapStart) && (ptrStart < ptrEnd))) ?
+			1 : 0;
+	/* clang-format on */
+}
+
+
+/* Ensure kernel won't make a bad access on a malformed elf during load. */
+static int process_loadValidateElf32(void *iehdr, size_t size)
```
Comment (agkaminski) [inline proc/process.c:413]: I get what you were trying to do, but we don't load anything here
```suggestion
static int process_validateElf32(void *iehdr, size_t size)
```

File: proc/process.c [hunk 6]
```diff
@@ -395,26 +395,185 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
...
+	}
+
+	return 0;
+}
+
+
 #ifndef NOMMU
 
+
+/* There are no NOMMU 64 bit targets.. */
```
Comment (agkaminski) [inline proc/process.c:486]: I don't think we need to justify placement after ifdef that hard ;)
```suggestion
```

**Review by agkaminski — COMMENTED** [review]
Some additional nitpicks, otherwise LGTM

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**